### PR TITLE
Add contributor guidelines and pre-commit hooks

### PR DESCRIPTION
**Pull request by tbennun**

### CODE REVIEW

**Review by bvanessen — CHANGES_REQUESTED** [review]
Need to finish the clang format first, but generally looks good.

**Comment by tbennun** [comment]
@benson31 @bvanessen with the new formatting merged, this is now ready for (re-)review.

I ran it on different systems and, since `pre-commit` hardcodes the Python binary path (and the same python binary points to different versions on different systems), I had to go to `.git/hooks/pre-commit` and modify the following line to:
```
INSTALL_PYTHON=`which python3`
```

This won't be an issue on most systems, it's just an unfortunate combination of two issues:
1. The same exact path points to two versions, and 
2. Python 3.7 doesn't install `pre-commit` properly because of an issue installing the latest version of PyYAML

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+## Setting up the repository
+
+After cloning LBANN, se
+Install `pre-commit` via the [instructions](https://pre-commit.com/#install) and set it up on the repository as follows:
+
+```sh
+/path/to/lbann $ pre-commit install
+```
+
+Make sure you have `clang-format` installed for C/C++ formatting and `yapf` for Python formatting.
```
Comment (bvanessen) [inline CONTRIBUTING.md:20]:
```suggestion
Make sure you have `clang-format` installed for C/C++ formatting and `yapf` for Python formatting.
Note that once `pre-commit` is installed in the `.git/hooks` directory of a repository, you will need to execute
all of your `git` commands from a system with a `python` that has `pre-commit` installed.
```

**Review by bvanessen — APPROVED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+If not ready, **rebase** the commits onto the latest commit. Avoid merge commits.
+
+## Style guidelines
+
+For C/C++ and GPU code, we follow the [LLVM coding style](https://llvm.org/docs/CodingStandards.html) with
+adaptations, see the [coding style README](https://github.com/LLNL/lbann/blob/develop/README_coding_style.txt) and the
+[clang-format configuration](https://github.com/LLNL/lbann/blob/develop/.clang-format) for more information.
+
+For Python code, we follow the [Google coding style](https://google.github.io/styleguide/pyguide.html) guidelines,
+but allow some exceptions to create layers in the LBANN Python frontend.
```
Comment (bvanessen) [inline CONTRIBUTING.md:34]:
```suggestion
but allow some exceptions to create layers in the LBANN Python frontend.

## Troubleshooting

When working on systems with older `python@3.7.2` there is a trick to getting pre-commit installed via `pip`:
`python3 -m pip install --user pyyaml==5.1`
`python3 -m pip install --user pre-commit`
```

**Comment by tbennun** [comment]
@bvanessen should work much smoother now

File: scripts/pre-commit-hook.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+CLANG_FORMAT="${CLANG_FORMAT:-clang-format}"
```
Comment (bvanessen) [inline scripts/pre-commit-hook.sh:3]:
```suggestion
CLANG_FORMAT="${CLANG_FORMAT:-$(which clang-format)}"
```

File: scripts/pre-commit-hook.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# Test for modified files
+echo -n "Testing modified files... "
+files=$(git diff --cached --name-only --diff-filter=ACM | grep -E "(pp|cu|proto)$")
+if [ "$files" == "" ]; then
+    echo "No files to test."
+    exit 0
+fi
+
+if ! [ -x "$CLANG_FORMAT" ]; then
+    echo "clang-format not found. Either install it or set the CLANG_FORMAT environment variable to its path"
```
Comment (bvanessen) [inline scripts/pre-commit-hook.sh:14]:
```suggestion
    echo "clang-format not found (CLANG_FORMAT=${CLANG_FORMAT}). Either install it or set the CLANG_FORMAT environment variable to its path"
```

**Review by bvanessen — APPROVED** [review]

**Pull request merged**